src/compiler/checker.ts

```diff
@@ -8823,26 +8832,62 @@ namespace ts {
             }
         }
 
-        // When intersecting unions of unit types we can simply intersect based on type identity.
-        // Here we remove all unions of unit types from the given list and replace them with a
-        // a single union containing an intersection of the unit types.
-        function intersectUnionsOfUnitTypes(types: Type[]) {
-            const unionIndex = findIndex(types, t => (t.flags & TypeFlags.UnionOfUnitTypes) !== 0);
-            const unionType = <UnionType>types[unionIndex];
-            let intersection = unionType.types;
-            let i = types.length - 1;
-            while (i > unionIndex) {
+        // Check that the given type has a match in every union. A given type is matched by
+        // an identical type, and a literal type is additionally matched by its corresponding
+        // primitive type.
+        function eachUnionContains(unionTypes: UnionType[], type: Type) {
+            for (const u of unionTypes) {
+                if (!containsType(u.types, type)) {
+                    const primitive = type.flags & TypeFlags.StringLiteral ? stringType :

```

Comment:
This logic is very close to `getBaseTypeOfLiteralType`; can you use that instead? That also covers enums, though I'm not sure if they come into play here.

Reply:
Although a 200 member enum could easily cause the same issues, so is _probably_ worth covering!

Reply:
No, we can't use that instead because it maps an enum literal type to its defining union type (the union of all enum literal types in the same enum type declaration) which we don't want here. We only want to handle the special "infinite sets" denoted by `string`, `number`, and `symbol`.